File: algorithm/198.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>
#define MAX(x, y) (x < y ? y : x)
int rob(int *nums, int numsSize)
{
    if (numsSize == 0 || nums == NULL)
    {
        return 0;
    }
    if (numsSize == 1)
    {
        return 1;
    }
    int *dp = (int *)calloc(numsSize, sizeof(int));
    dp[0] = nums[0];
    dp[1] = MAX(nums[0], nums[1]);
    for (int i = 2; i < numsSize; i++)
    {
        dp[i] = MAX(dp[i - 1], dp[i - 2] + nums[i]);
    }
    return dp[numsSize - 1];
}

int main(int argc, char *argv[])
{
    int a[4] = {1, 2, 3, 1};
    printf("%d\n", rob(a, 4));
    return 0;
}

File: go/goalgor/algor2/198.go
package main

func Max(i, j int) int {
	if i < j {
		return j
	}
	return i
}

func rob(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	if len(nums) == 2 {
		return Max(nums[0], nums[1])
	}
	record := make([]int, len(nums)+1)
	record[0] = nums[0]
	record[1] = Max(nums[0], nums[1])
	for i := 2; i < len(nums); i++ {
		record[i] = Max(record[i-1], record[i-2]+nums[i])
	}
	return record[len(nums)-1]
}

// func main() {
// 	nums := []int{1, 2, 3, 1}
// 	fmt.Println(rob(nums))
// }
